# Modification of Smith-Waterman Algorithm for Identifying Potential miRNA Binding Sites

MicroRNAS (miRNAs) are short non-coding RNAs, about 21 to 25 nucleotides in length, found in eukaryotic genomes which act as post-transcriptional regulators of gene expression. They recognize and bind to a complementary sequence usually in the 3’ UTR of a target mRNA, and this binding has an inhibitory effect, resulting in decreased translation of the target mRNA. The discovery of miRNAs is of interest because they act as another mechanism of gene regulation in addition to directly modifying the genome, and miRNA levels have also recently been linked to the development of cancer. Our program uses a modified Smith-Waterman algorithm to identify ideal binding sites given a miRNA and a 3’ UTR of some target mRNA.

## Introduction 
The general mechanism of miRNAs involves export to the cytoplasm for incorporation into the RNA-induced silencing complex (RISC), and these RISCs can then target the 3’UTRs of mRNAs for inhibition (Meltzer, 2005). The base pairing mechanism is what gives miRNAs specificity in inhibition, and the degree of mismatch between the miRNA and target mRNA is a determining factor in whether the target mRNA is inhibited or silenced. Because miRNAs may target multiple genes, and several miRNAs may regulate a single target, there is still a lot more to be discovered about the interplay between mRNA and miRNA in gene expression, and its role in disease.

One example of potential use of miRNAs is in cancer diagnosis: tumors derived from tissues with a common embryonic origin share similar miRNA expression patterns, so miRNA expression data can be useful for determining the origins of poorly defined cancers and in designing individualized cancer therapy (Meltzer, 2005). Expression of miRNAs appears to be lower in cancer when compared to normal tissue, which is consistent with the idea that miRNAs act as negative regulators and may be tumor suppressors. However, a recent study by He et al. (2005) found that overexpression of miRNAs in the c13orf25 cluster appears to be related to a quicker onset of leukemia development and an increase in the frequency of cancer development.  Thus, there appears to be a dual role for miRNA: in one context, the c13orf25 cluster appears to act as an oncogene; in another, it seems to antagonize the effects of other oncogenes. 

In order to further understand the interactions between miRNAs and gene expression as related to cancer prognosis, it would be helpful to identify candidate miRNA target genes and sequences through bioinformatics approaches. This information could help serve as a screening to determine whether a certain miRNA sequence is able to bind to and inhibit translation of a library of mRNAs. In addition, introducing non-native miRNAs to where they have a complementary binding site in the 3’UTR of a gene could serve as an alternative method of gene knockdown. Understanding the specific miRNA targets on a known inhibited sequence could also provide insight on the range of binding sites and binding affinities that are satisfactory for inhibition; and these sites could perhaps be introduced into another mRNA for downregulation by miRNAs. This bioinformatics approach may provide insight into experimental design, and could be paired with in-vitro methods of introducing miRNA into cells to determine if computationally identified binding targets are in fact correlated with reduced expression.


## Results and Conclusions
We developed a modified version of the Smith-Waterman algorithm to identify ideal binding sites given an miRNA and a 3’UTR mRNA sequence. The Smith-Waterman algorithm is a program designed to find the best local alignment of two sequences and uses a dynamic programming approach to recursively compute the best alignment of all possible subsequences. Although Smith-Waterman is typically used to identify similar sequences, we can modify its design to instead identify two compatible complementary sequences. The given program takes a query (the short miRNA sequence) as a command line argument and a target input (the 3’UTR sequence for the gene of interest) and runs Smith-Waterman after reversing the query to find and score complementary matches. On a high level, the algorithm’s recurrence looks for the best scoring to fill out a scoring matrix, and uses a traceback to identify the best local alignment. The highest scoring matches are then returned as the output.

The first modification made to the algorithm was adding weighting for matches in the first ~7 nucleotide seed region in the query miRNA. Grimson et al. (2007) found that the most downregulation after miRNA introduction occurred when there was high complementary pairing of the seed region to the mRNA; this seed includes a 6mer (miRNA nt 2-7), two 7mers (miRNA A1-7 or 2-8), and one 8mer (miRNA 1-8). Since the seed region appears highly important for repression due to miRNA binding, we added differential weightings to favor each of these matches. The 6mer is the perfect 6 nucleotide match to the miRNA seed, and we used this as the highest weighted region because the 7mer and 8mer both depend on perfect matching of this 6mer region. We also added additional weightings for the two 7mer matches: if the first nucleotide position included an A, or if the 8th nucleotide matched that of the miRNA; and also a weighting for the 8mer if all eight sites matched. In modifying the weighting as the algorithm iterates over the miRNA sequence and some region on the 3’UTR, the recurrence will preferentially choose to continue matches if the first couple of nucleotides match the seed, and these will be highest weighted when calculating the traceback.

Another metric for consequential miRNA pairing which may be linked to enhanced repression is Watson-Crick pairing at nucleotides 12-17 miRNA; especially at nucleotides 12-16, was most associated with downregulation (Grimson et al., 2007). To incorporate ideal Watson-Crick pairing at nucleotides 12-17 in the miRNA into the program, we used an alternative approach by penalizing mismatches more severely in this region. Sites with a “similar” match were instead scored as “mismatches,” such that most sequences would be heavily penalized for lacking matches in the 12-17 region; but sites with ideal Watson-Crick pairing would not be penalized and thus have an elevated score compared to other pairings.

Effective miRNA sites preferentially reside near both ends of the 3’UTR; but sites within 15 nucleotides of the stop codon were found to be less effective when compared to sites elsewhere in the 3’UTR (Grimson et al., 2007). To account for this finding, matches scored in the first and last quartiles of the target mRNA sequence were boosted by a small weighting (0.5x) to promote traceback of sequences in these areas. A large negative weight was added (-2.0x) to heavily disfavor any potential binding sites considered within the first 15 nucleotides of the 3’UTR. 

Finally, proximity of the miRNA binding sites within 100 nt to each other was found to act synergistically, so we added a modification to boost the scoring of sites that were closely spaced (Grimson et al., 2007). This modification could most likely be improved, as several strategies were used to increase running time efficiency at the slight cost of accuracy.  Because looping through all the potential tracebacks for 100 nt proximity would take at minimum O(n2) running time, this is extremely slow considering the size of the input. We instead assigned a minimum length (~20 nt) and minimum score (~35) for which it would be efficient to check for score boosting; otherwise, sites below these parameters were ignored to preserve running time efficiency. It was difficult to gauge what values for these parameters would best effectively represent the synergistic effects of closely spaced sites; so we relied on a validation program to check the outputs.

We ran a validation program to compare three different outputs: the original Smith-Waterman algorithm, SW1 (weighting added to the seed region and for Watson-Crick pairing at nt 12-17), and SW2 (same modifications as SW1, but considering proximity of matches to the UTR ends and closely spaced sites). Ideally, SW2 would be the best-performing algorithm as it incorporates all the biological parameters as identified above. The first parameter checked was on the lengths of matches found by each algorithm (Figure 1). The input miRNA query was 22 nt in length, so no perfect match was found in the 3’UTR; but all three iterations of the program found matches within 19-21 nucleotides in length. Both SW1 and SW2 seemed to introduce slightly more gaps compared to the original program; this result may be due to the two algorithms weighing other factors higher than simply a perfect complementary match. Compared to SW1, the modified SW2 program appears to minimize gaps within the miRNA in the highest scoring matches (Figure 2).

Another parameter to be validated was the degree of similarity in the seed region in the matches found by the original, compared to SW1 and SW2. Although the original algorithm found seeds of the highest (~67%) and lowest percent identity (~21%), SW1 and SW2 had more concentrated % identity matches within the 30% to 55% identity range (Figure 3). Thus, our algorithm appeared to at least remove from consideration any matches with a lower than 30% identity in the seed region. However, this algorithm could be improved since it appears to somehow miss the match of highest % identity that was identified by the original program.

Another parameter of interest was the degree of similarity in the Watson-Crick pairing 12-17 nucleotide region. The SW1 algorithm performed best in finding the highest degree of %identity matches in this 12-17 nt length; and both SW1 and SW2 performed better than the unaltered algorithm in % identity matches (Figure 6). SW2 seemed to find matches only in the range of 40-70% identity in the 12-17 nt region, and the original algorithm found regions of 17-65% identity instead so there was a marked improvement in the % identity of this region. However, the original algorithm appeared to be better suited for identifying higher % similarity; but since good matching of this region appears to be linked to the increased suppression effects of miRNA binding, it is better to promote higher % identity. Furthermore, the SW2 algorithm clearly finds more matches within the first and last quartiles of the 3’UTR; and when compared to the original algorithm, almost no matches are found within the middle 50 percent of the sequence (Figure 10). Further validation plots can be found in the attached figures.

Overall, our modified Smith-Waterman algorithm performed marginally better in finding matches at the seed and 12-17nt regions; it also prioritized matches found within the first and last quartiles of the 3’UTR as intended. However, it was difficult to determine whether the highest scoring matches as determined by SW2 were indeed more closely spaced. One drawback we ran into while implementing this last modification was creating a balance between introducing too many gaps between certain sites proximal to the 3’UTR ends, and finding closely spaced sites to each other without creating too much of a bias for sites that were that concentrated at the ends of the 3’UTR region. The similarity of the SW1 and SW2 output validates that this weighting did not severely bias the results; but without wet lab experimentation it is difficult to determine how much of a bias is optimal. Since the algorithm incorporated many different biological considerations, it was also a challenge to identify which factors should have which weightings associated to them in the context of all the factors. In the future, an additional aspect to implement could be to also take into consideration the AU-rich context of the sites as this was another aspect of increased miRNA suppression functionality (Grimson et al., 2007). It would also be interesting to test if a high-scoring match between miRNA and 3’UTR as determined by this algorithm could in fact have an inhibitory effect when tested experimentally. Another method to perhaps improve the functionality would be to run an iteration over the validation program, and to choose optimal values for each of the parameters (i.e. gap penalty, seed 6mer match, etc). There are many potential future iterations on this program, and the modularity remains important as new findings could be used to better inform identification of complementary miRNA binding sites.


## References

Grimson, et al., “MicroRNA targeting specificity in mammals: Determinants beyond
seed-pairing” Mol. Cell 27:91-105 (2007)

He, et al.,  “A microRNA polycistron as a potential human oncogene” Nature 435: 828-833 
(2005)

Johnson, et al., “RAS is regulated by the let-7 microRNA family” Cell 120: 635-647 (2005) 

Meltzer, “Small RNAs with big impacts” Nature 435: 745-746 (2005)


## Figures
Original SW is the given Smith-Waterman algorithm; altered SW1 is the program with seed and Watson-Crick pairing. Altered SW2 contains the same changes as SW1 but with proximity to UTR ends and closely spaced sites in consideration. This naming convention is preserved in the following figures. 

<img width="332" alt="Screenshot 2022-11-15 at 11 55 33 PM" src="https://user-images.githubusercontent.com/55102022/202087310-2705859e-0188-4399-bd07-15fb3a48ff19.png">
<img width="338" alt="Screenshot 2022-11-15 at 11 56 48 PM" src="https://user-images.githubusercontent.com/55102022/202087440-6156e139-1202-412b-b6a2-8a99e6777d05.png">
<img width="361" alt="Screenshot 2022-11-15 at 11 56 59 PM" src="https://user-images.githubusercontent.com/55102022/202087468-ae7bbf79-6193-4da7-b508-a3940cb167d5.png">

